02cpp1/sec03Pointers.md

```diff
@@ -0,0 +1,364 @@
+```cpp
+int addWithConstPointers(int const * const a, int const * const b)
+{
+    return (*a) + (*b);
+}
+```
+
+## Using libraries which use raw pointers
+
+Some commonly used external libraries like LAPACK or GNU Science Library, which are C compatible, will require you to pass raw pointers to functions. In this case, you do not have to forego using smart pointers in your own code. A smart pointer is a wrapper for a raw pointer, and the raw pointer can be retrieved using `.get()` on a smart pointer. You can then pass this raw pointer to the library function. 
+
+These functions will usually use C-style arrays, which are pointers declared and used in a function in the following way:
+```cpp
+// N by N float matrix
+int N = 1000;
+float *myFloatMatrix = new float[N * N]
+int res = myMethod(myFloatMatrix);
+```
+
+To use a smart point we could write:
+```cpp
+int N = 1000;
+std::unique_ptr<float[]> myFloatMatrix = std::make_unique<float[]>(N*N);
+int res = myMethod(myFloatMatrix.get());
+```
+In general we will not focus on libraries of this kind, but it is good to be aware of them as you may come across them in the future when working with academic software. 
+
+## Smart Pointers, Performance, and Optimisation
+
+Smart pointers are more complex than raw pointers, and can incur some overheads. Shared pointers require some additional memory in order to perform the reference counting. The compiler can automatically optimise away almost all of the time overhead associated with shared or unique pointers, but without optimisation they can be significantly slower. We'll discuss how to perform this kind of optimisation at compile time later on when we discuss programming for high performance.
+
+Accessing weak pointers will remain slow even with optimisation turned on; this is because of the `.lock()` operation, which has to make a copy of an existing shared pointer before it can be dereferenced. If accessing pointers is a bottleneck in your program (for example a graph traversal where nodes are linked by pointers), then you may want to consider replacing the weak (and therefore non-owning) pointers with raw pointers if you can ensure that they won't attempt to dereference invalid memory. 
+
+## Memory Problems
+
+All kinds of pointers can cause some memory problems through improper handling. Whenever you use pointers, make sure that you're careful, understand when your memory is being allocated and deallocated, test your functions thoroughly for memory access issues and use a profiler to check for memory leaks. We'll discuss profiling memory usage using `valgrind` in a later lecture. 
```

Comment:
Can there be a summary section here? The pointer material is perfectly detailed IMO but it feels like there are a few key takeaways that could be emphasised.

Could also collate all the good practices you've listed throughout the page